src/seq/relics.rs

```diff
@@ -0,0 +1,131 @@
+use crate::control::SosAction;
+use crate::memory::title_sequence_manager::{RelicButton, TitleSequenceManagerData};
+use crate::state::GameState;
+use joystick::prelude::*;
+use log::{info, warn};
+use seq::prelude::*;
+
+use crate::seq::button::ButtonPress;
+
+#[derive(Default)]
+enum RelicScreenFSM {
+    #[default]
+    Eval,
+    PressButton,
+}
+
+#[derive(Default)]
+pub struct SeqRelicList {
+    fsm: RelicScreenFSM,
+    btn: ButtonPress,
+    first: Option<String>,
+    prev: Option<String>,
+    done: bool,
+}
+
+impl SeqRelicList {
+    pub fn create() -> Box<Self> {
+        Box::new(Self::default())
+    }
+
+    // Shorthand to get the currently selected relic
+    fn get_current_relic<'a>(&self, tsmd: &'a TitleSequenceManagerData) -> Option<&'a RelicButton> {
+        tsmd.relic_buttons
+            .items
+            .iter()
+            .find(|&button| button.selected)
+    }
+
+    // Set up FSM to move to next relic in the list
+    fn next_relic(&mut self) {
+        self.fsm = RelicScreenFSM::PressButton;
+        self.btn = ButtonPress {
+            action: SosAction::MenuDown,
+            ..Default::default()
+        };
+    }
+
+    // Set up FSM to toggle relic state
+    fn toggle_relic(&mut self) {
+        self.fsm = RelicScreenFSM::PressButton;
+        self.btn = ButtonPress {
+            action: SosAction::Confirm,
+            ..Default::default()
+        };
+    }
+
+    // When done, press start
+    fn press_start(&mut self) {
+        self.done = true;
+        self.fsm = RelicScreenFSM::PressButton;
+        self.btn = ButtonPress {
+            action: SosAction::Start,
+            ..Default::default()
+        };
+    }
+
+    // Check if we have wrapped by comparing current selected relic to the first in the list.
+    // If there is no first relic, store its name now. We must have moved from a different
+    // relic before (a previous relic name is held and updated).
+    fn check_done(&mut self, cur_relic: &str) -> bool {
+        if let Some(prev_name) = &self.prev {
+            if let Some(first_name) = &self.first {
+                let wrapped = cur_relic == first_name && cur_relic != prev_name;
+                self.prev = Some(cur_relic.to_owned());
+                return wrapped;
+            }
+        } else {
+            self.prev = Some(cur_relic.to_owned());
+            self.first = Some(cur_relic.to_owned());
+        }
+        false
+    }
+}
+
```

Comment:
this is alot more succinct than what I did in the python tas;
I just checked the entire list to see if it matched on every action :)

I like this version better because you dont need a predetermined state to compare to.

Reply:
Yeah I kinda struggled back and forth with it until settling for this one. My first stab did something similar.